Add support for authorized, private registries

rexxars opened the pull request:
This PR adds support for private registries, and also handles authorization against these.

Fixes #18.

sindresorhus commented on index.js:
```diff
@@ -1,28 +1,19 @@
 var got = require('got');
-var registryUrl = require('registry-url');
-var rc = require('rc');
+var registryUrl = require('registry-auth-token/registry-url');
```
Why change this? Seems like it's just a copy of `registry-url`.

rexxars replied:
I originally added the registry-url code to the auth token repo because a PR I sent to a different project ([ied](https://github.com/alexanderGugel/ied) wanted to reduce the number of dependencies they added. I'd be fine with using the `registry-url` module if you prefer the extra dependency.

sindresorhus replied:
> I'd be fine with using the registry-url module if you prefer the extra dependency.

Yes

sindresorhus commented on index.js:
```diff
@@ -1,28 +1,19 @@
 	var headers = {};
 
-	if (token) {
-		if (process.env.NPM_TOKEN) {
```
Doesn't look like your module covers this use-case?

rexxars replied:
Actually, [it does](https://github.com/rexxars/registry-auth-token/blob/master/index.js#L65-L67). See [tests](https://github.com/rexxars/registry-auth-token/blob/master/test/auth-token.test.js#L71-L103).

sindresorhus replied:
I see. I thought it had to be `NPM_TOKEN`.

sindresorhus commented:
Ensure it still supports private scoped packages. Also make sure you handle these: https://github.com/sindresorhus/package-json/pull/27 https://github.com/sindresorhus/package-json/pull/23 https://github.com/sindresorhus/package-json/pull/14

sindresorhus commented:
We should probably add to the readme what kind of authentication is supported:

- Public packages
- Private scoped packages
- Basic auth private packages
- Private registries

rexxars commented:
Updated based on feedback, commits squashed.

> Ensure it still supports private scoped packages

Private scoped packages are just using the bearer tokens, so those work fine

> Also make sure you handle these: #27 #23 #14

- #27 - No test for this one, so I added one. Slightly ugly, but couldn't figure out a better way to check if any authorization header was being sent.
- #23 - This PR was closed without being merged, and seems to address `npm@1.x`, which I personally don't think is worth pursuing
- #14 - As mentioned above, this is taken care of by [registry-auth-token](https://github.com/rexxars/registry-auth-token)

> We should probably add to the readme what kind of authentication is supported:
> - Public packages
> - Private scoped packages
> - Basic auth private packages
> - Private registries

Authorization-wise, we just need to support bearer and basic tokens stored in npmrc, which we do (tests for both cases are present). Secondly we need to resolve any scope present to a registry URL, which the `registry-url` module handles. I'll leave it up to you to add this info to the readme.

I've also updated the registry mock module I wrote to be compatible down to Node 0.10, so the tests now pass on all supported versions. Let me know if you want me to do anything else with this PR.

rexxars commented:
Added a note about which authentication types are supported to the readme, let me know if you want anything else changed.

rexxars commented:
This will allow [yeoman/update-notifier](https://github.com/yeoman/update-notifier) to work with private registries, by the way \o/

sindresorhus commented:
Super work @rexxars! ✨

Invited you to the repo since this is an [open open source](http://openopensource.org) project :) If you're not interested, just decline. No worries.

rexxars commented:
Thanks :+1: Accepted :-)
